src/Microsoft.AspNet.Server.Kestrel/Http/FrameOfT.cs

```diff
@@ -93,6 +97,9 @@ public class Frame<TContext> : Frame
                                 await FireOnStarting();
                             }
 
+                            _requestBody.PauseAcceptingReads();
+                            _responseBody.PauseAcceptingWrites();
+
                             if (_onCompleted != null)
                             {
                                 await FireOnCompleted();
```

Comment:
If it throws, you move into the finally block and shutdown the connection and throw away the streams with it - they are only reused over the connection.